### ISSUE TITLE
Building from source

### ISSUE THREAD

**Issue by selfisekai** [Flutter-Foundation/nest#1]
I think a lot of missed potential in Flutter is suitability to Linux distros: possibility to package Dart and Flutter themself, allowing to package apps created with them. Is this something this project would accept contributions for?

(I'm the maintainer of flutter and dart packages at Alpine Linux)

**Comment by matthew-carroll** [comment]
We're certainly happy to consider it. We're here to deliver value where the community wants it. The first thing we need to do is get the engine building, uploaded, and delivered through Flock so that we can match baseline Flutter. From there, we'd be happy to see where we can take things.

**Comment by RossComputerGuy** [comment]
I'm interested in this too, maintainer for Flutter in Nixpkgs. I got the engine building in nix but it's a bit of a pain.

**Comment by matthew-carroll** [comment]
Can you guys elaborate on what "Build from source" means? If this were implemented, how would it change Flutter? What would it look like to use this capability?

**Comment by RossComputerGuy** [comment]
The TLDR of it is Flutter fetches many prebuilt components, the engine for instance. These prebuilts can produce impurities in certain environments, especially in the Nix ecosystem.

The other side of it is the engine, the engine uses gclient to fetch the source. The sources contains forks or original sources of libraries and tools packagers and developers already have. They also contain prebuilt binaries. The prebuilt binaries inflates the fetch size and causes issues outside of FHS environments (Nix & likely Guix).

Relying on the developer to install packages which they already will likely have (gtk, clang, ninja, etc) can decrease time to get the source code, reduce the burden of maintaince, and just make the overall experience better.

**Comment by RossComputerGuy** [comment]
See https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/flutter/engine/source.nix for how I was able to produce a fully pure and reproducible source. As you can see, a lot has to be removed. Hooks have to be disabled in order to not run tools which will fail due to Nix not using FHS.

There's also all of this post unpack work inside the actual engine build:

https://github.com/NixOS/nixpkgs/blob/95c1043ff5531b4348c50fa92763954df98399e/pkgs/development/compilers/flutter/engine/package.nix#L194-L244

It's a lot to maintain but makes it possible. I also recently might've ran into a problem where 3.26 changes something and is producing unreproducible sources which makes fetching the source undeterministic.

**Comment by matthew-carroll** [comment]
CC @jezell

**Comment by jezell** [comment]
I really like the idea of enabling some more customization of the lower level parts of the engine. I had similar challenges making custom canvaskit builds on web. There's a lot of potential to unlock in customizing some of the lower level parts of the engine build. The key thing to figure out is how to do this in a way that doesn't conflict too much with changes that are being auto merged, but what you accomplished there is proof that it can be done.
